python_transport/wirepas_gateway/keep_alive_service/keep_alive_service.py

```diff
@@ -0,0 +1,245 @@
+
+
+class KeepAliveServiceThread(Thread):
+    def __init__(self, sink_manager, mqtt_wrapper,
+                 keep_alive_interval_s=300,
+                 keep_alive_timezone_offset_mn=0):
+        """ Thread sending periodically keep alive messages to the network.
+
+        Args:
+            sink_manager: The sink manager to send sinks the keep alive messages.
+            mqtt_wrapper: The mqtt wrapper to get access to queue level of the mqtt broker.
+            keep_alive_interval_s (int): Default to 300 seconds.
+                The interval in seconds between keep-alive messages.
+            keep_alive_timezone_offset_mn (int): Default to 0.
+                Time zone offset from UTC in minutes (-840 to +720) of the keep alive message.
+        """
+        Thread.__init__(self)
+
+        # Daemonize thread to exit with full process
+        self.daemon = True
+
+        self.sink_manager = sink_manager
+        self.mqtt_wrapper = mqtt_wrapper
+
+        self.disconnected_sinks = set()  # All sinks that are detected as disconnected to the gateway
+        self.connected_sinks = set()  #  All sinks that are detected as connected to the gateway
+
+        self.keep_alive_interval_s = keep_alive_interval_s
+        self.keep_alive_timezone_offset_mn = keep_alive_timezone_offset_mn
+
+    def prepare_keep_alive_msg(self):
+        """ Prepare and return a keep alive message. """
+        rtc_timestamp_ms = None
+        time_zone_offset = None
+
+        rtc_timestamp_ms = int(time() * 1000)
+        time_zone_offset = self.keep_alive_timezone_offset_mn
+
+        gateway_status = int(self.mqtt_wrapper.connected)
+        keep_alive_msg = KeepAliveMessage(KEEP_ALIVE_SERVICE_VERSION,
+                                          gateway_status,
+                                          rtc_timestamp_ms,
+                                          time_zone_offset,
+                                          self.keep_alive_interval_s)
+
+        return keep_alive_msg
+
+    def send_keep_alive_msg_to_sink(self, sink) -> bool:
+        """
+        Send the keep alive message to the network.
+        Returns True if the keep alive message could be sent to the sink, False otherwise.
+
+        Args:
+            sink: Sink to send the keep alive message to.
+        """
+        retries_left = 3
+        res = wmm.GatewayResultCode.GW_RES_UNKNOWN_ERROR
+
+        while retries_left > 0 and res != wmm.GatewayResultCode.GW_RES_OK:
+            retries_left -= 1
+            keep_alive_message = self.prepare_keep_alive_msg()
+            payload = keep_alive_message.encode_tlv()
+            logging.debug("Send the following keep alive payload to sink %s: %s",
+                          sink.sink_id, payload.hex())
+
+            res = sink.send_data(
+                dst=BROADCAST_ADDRESS,
+                src_ep=KEEP_ALIVE_SRC_EP,
+                dst_ep=KEEP_ALIVE_DST_EP,
+                qos=0,
+                initial_time=0,
```

Comment:
If I recall right, it is already the default value, so could be skipped

Reply:
This argument is required by the method. I don't think there is any default value for the initial_time.